core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java

```diff
@@ -0,0 +1,537 @@
+    * @throws NullPointerException if {@code lockOwner} is {@code null}.
+    */
+   public void release(Object lockOwner) {
+      Objects.requireNonNull(lockOwner, "Lock Owner should be non-null");
+
+      if (trace) {
+         log.tracef("Release lock for %s.", lockOwner);
+      }
+
+      LockPlaceHolder wantToRelease = lockOwners.get(lockOwner);
+      if (wantToRelease == null) {
+         if (trace) {
+            log.tracef("%s not found!", lockOwner);
+         }
+         //nothing to release
+         return;
+      }
+
+      final boolean released = wantToRelease.setReleased();
+      if (trace) {
+         log.tracef("Release lock for %s? %s", wantToRelease, released);
+      }
+
+      LockPlaceHolder currentLocked = current;
+      if (currentLocked == wantToRelease) {
+         if (!casRelease(currentLocked)) {
+            if (trace) {
+               log.tracef("Releasing lock for %s. It is the current lock owner but another thread changed it.", lockOwner);
+            }
+            //another thread already released!
+         } else {
+            tryAcquire();
+         }
+      }
+   }
+
+   /**
+    * @return {@code true} if this lock has no lock owner pending.
+    */
+   public boolean isEmpty() {
+      return isFree() && lockOwners.isEmpty();
+   }
+
+   /**
+    * @return {@code true} if the lock is not acquired and no lock owners are queued.
+    */
+   public boolean isFree() {
+      return current == null && pendingRequest.isEmpty();
+   }
+
+   /**
+    * @return the current lock owner or {@code null} if it is not acquired.
+    */
+   public Object getLockOwner() {
+      LockPlaceHolder lockPlaceHolder = current;
+      return lockPlaceHolder == null ? null : lockPlaceHolder.owner;
+   }
+
+   /**
+    * It checks if the lock is acquired.
+    * <p/>
+    * A {@code false} return value does not mean the lock is free since it may have queued lock owners.
+    *
+    * @return {@code true} if the lock is acquired.
+    */
+   public boolean isLocked() {
+      return current != null;
+   }
+
+   /**
+    * It forces a deadlock checking.
+    */
+   public void deadlockCheck() {
+      LockPlaceHolder holder = current;
+      if (holder != null) {
+         for (LockPlaceHolder pending : pendingRequest) {
+            pending.checkDeadlock(holder.owner);
+         }
+      }
+   }
+
+   /**
+    * It tests if the lock has the lock owner.
+    * <p/>
+    * It return {@code true} if the lock owner is the current lock owner or it in the queue.
+    *
+    * @param lockOwner the lock owner to test.
+    * @return {@code true} if it contains the lock owner.
+    */
+   public boolean containsLockOwner(Object lockOwner) {
+      return lockOwners.containsKey(lockOwner);
+   }
+
+   private void onCanceled(LockPlaceHolder canceled) {
+      if (trace) {
+         log.tracef("Release lock for %s. It was canceled.", canceled.owner);
+      }
+      LockPlaceHolder wantToRelease = lockOwners.get(canceled.owner);
```

Comment:
no they are always the same. 
the clear from the map occurs when lock() is invoked (and it throws timeout or deadlock). When the exception is thrown, I'm not sure if release() may never be invoked so I perform the cleanup. 
the second scenario is when the release is invoked (normal scenario, lock is acquired and then released)